# Assignment 7: Image Processing w/ CNN - Oscar Hernandez 

## Purpose

We are advising the management team of a website provider on which machine learning model works best for image classification. They are looking for advise on a tool that can be used to automatically label images that end users provide. This report will outline the work that went into building several binary classification models while using a 2x2 completely crossed experimental design. Specifically, we will be training/testing convolutional neural networks (CNNs) using Keras running on top of TensorFlow. We will be utilizing a subset of the Asirra data set which includes images of dogs and cats to develop our CNNs. Our main objective/concern is developing a classification model with the highest possible accuracy.  

#### This report will be broken up into Sections that cover the specific methodology that went into arriving at the final recommendation. The report will include and/or cover, at minimum, the following specific items and/or tasks:
* Complete a quick exploration of the dogs and cats data set 
* Partitioning of the data set into training and test sets (due to size of data set, we chose not to create a validation set)
* Train four CNNs using identical hyperparameters except the number of layers and number of neurons within each layer
* Check each CNN for accuracy using our training and test subset
* Evaluate the differences (focus on accurracy) observed across treatment combinations  
* Provide final management recommendation 


### Section 1 - Exploratory Data Analysis & Preparation

#### Section 1 covers loading the data sets that will be used to train our CNNs. We will also create the y labels for our data which were not provided to us.   

In [74]:
#Load all the necessary packages we need to complete the exercise

import os
import pandas as pd
import numpy as np
import time
import nbconvert
from sklearn.model_selection import train_test_split
from keras.models import Sequential
from keras.layers import Dense, Conv2D, Flatten

In [2]:
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'

In [3]:
import tensorflow as tf

In [45]:
#Import the cat/dog data as NumPY arrays
#We were provided two separate NumPY array files for cats and dogs
cats_1000_64_64_1 = np.load('cats_dogs_64-128/cats_1000_64_64_1.npy')
dogs_1000_64_64_1 = np.load('cats_dogs_64-128/dogs_1000_64_64_1.npy')

In [46]:
#We can infer from the output of shape attribute, that the cats data 
#is 1000 observations and each image is 64x64 pixels 
#This same info applies for the dogs data  
cats_1000_64_64_1.shape

(1000, 64, 64, 1)

In [47]:
#Now, let's combine the separate files into one NumPY array
#The first line of code stacks the files on top of each other
#along the column axis 
X_cat_dog = np.concatenate((cats_1000_64_64_1, dogs_1000_64_64_1), axis = 0) 

In [51]:
#Check the shape just to be sure 
X_cat_dog.shape

(2000, 64, 64, 1)

In [25]:
#Next we need to create our y labels using the concatenate function 
#where the first 1000 observaations are 0, which are the cat labels
#and the last 1000 observations are 1, which are the dog labels
#In our binary model, the 1 label means the observation is a dog
#and the 0 label means the observation is a cat 
y_cat_dog = np.concatenate((np.zeros((1000), dtype = np.int32), 
                      np.ones((1000), dtype = np.int32)), axis = 0)

In [52]:
#Finally, lets use the Scikit-Learn function to split both of 
#the 2D NumPY arrays into training and test sets
X_train, X_test, y_train, y_test = train_test_split(X_cat_dog, y_cat_dog,
                                                   test_size = 0.2, random_state = 7)

###### Key Takeaways 
* In this section, we started by loading the cat and dog data which were stored as numpy arrays
* We checked the shape of the arrays which is necessary since it is an input into our CNN later on
* We combined both arrays into one and then also created y labels 
* The final step was creating a training and test data set by using a 80/20 split 

### Section 2 - Build Convolutional Neural Networks

#### Section 2 will cover the steps conducted to create four CNNs. The only difference between each CNN is the number of layers and neurons per layer. This was done as part of the experiment to see the impact the change in these factors has on the accuracy of each CNN. 

#### We will be using Keras to build out each CNN. Also, we will be recording the time it takes to develop each one. 

#####  Classifier 1 Specs
* Two convolution layers
* 32 nodes per layer
* ReLU activation function except for the output layer (Sigmoid)
* Kernel size = 3 (3x3 filter matrix) 

In [62]:
#Each model will be built in the following way: 
#Create initial model object using Keras Sequential function
#Use the add method to create each layer 
#Compile the model by choosing the optimizer and cost function 
#Train the model using the cat/dog data set 
cnn1_start_time =time.time()

cnn1 = Sequential()
cnn1.add(Conv2D(32, kernel_size = 3, activation ="relu", input_shape=(64,64,1)))
cnn1.add(Conv2D(32, kernel_size =3, activation ="relu"))
#Connection between convolution and dense layer
cnn1.add(Flatten())
cnn1.add(Dense(1, activation = "sigmoid"))

cnn1.compile(optimizer = "rmsprop",loss="binary_crossentropy", 
             metrics=["accuracy"])

cnn1.fit(X_train, y_train, epochs = 5, batch_size = 20)

cnn1_elapsed_time = time.time() - cnn1_start_time

Epoch 1/5
1600/1600 [==============================] - 25s 16ms/step - loss: 7.9698 - acc: 0.5006
Epoch 2/5
1600/1600 [==============================] - 26s 16ms/step - loss: 8.0389 - acc: 0.5013
Epoch 3/5
1600/1600 [==============================] - 25s 16ms/step - loss: 8.0389 - acc: 0.5013
Epoch 4/5
1600/1600 [==============================] - 24s 15ms/step - loss: 8.0389 - acc: 0.5013
Epoch 5/5
1600/1600 [==============================] - 26s 16ms/step - loss: 8.0389 - acc: 0.5013


In [63]:
print (round(cnn1_elapsed_time,2), "seconds")

127.03 seconds


#####  Classifier 2 Specs
* Two convolution layers
* 16 nodes per layer
* ReLU activation function except for the output layer (Sigmoid)
* Kernel size = 3 (3x3 filter matrix) 

In [64]:
cnn2_start_time =time.time()

cnn2 = Sequential()
cnn2.add(Conv2D(16, kernel_size = 3, activation ="relu", input_shape=(64,64,1)))
cnn2.add(Conv2D(16, kernel_size =3, activation ="relu"))
#Connection between convolution and dense layer
cnn2.add(Flatten())
cnn2.add(Dense(1, activation = "sigmoid"))

cnn2.compile(optimizer = "rmsprop",loss="binary_crossentropy", 
             metrics=["accuracy"])

cnn2.fit(X_train, y_train, epochs = 5, batch_size = 20)

cnn2_elapsed_time = time.time() - cnn2_start_time

Epoch 1/5
1600/1600 [==============================] - 13s 8ms/step - loss: 8.0427 - acc: 0.5000
Epoch 2/5
1600/1600 [==============================] - 12s 7ms/step - loss: 8.0389 - acc: 0.5012
Epoch 3/5
1600/1600 [==============================] - 12s 8ms/step - loss: 8.0389 - acc: 0.5013
Epoch 4/5
1600/1600 [==============================] - 11s 7ms/step - loss: 8.0389 - acc: 0.5013
Epoch 5/5
1600/1600 [==============================] - 14s 9ms/step - loss: 8.0389 - acc: 0.5013


In [65]:
print (round(cnn2_elapsed_time,2), "seconds")

63.84 seconds


#####  Classifier 3 Specs
* Four convolution layers
* 32 nodes per layer
* ReLU activation function except for the output layer (Sigmoid)
* Kernel size = 3 (3x3 filter matrix) 

In [66]:
cnn3_start_time =time.time()

cnn3 = Sequential()
cnn3.add(Conv2D(32, kernel_size = 3, activation ="relu", input_shape=(64,64,1)))
cnn3.add(Conv2D(32, kernel_size =3, activation ="relu"))
cnn3.add(Conv2D(32, kernel_size =3, activation ="relu"))
cnn3.add(Conv2D(32, kernel_size =3, activation ="relu"))
#Connection between convolution and dense layer
cnn3.add(Flatten())
cnn3.add(Dense(1, activation = "sigmoid"))

cnn3.compile(optimizer = "rmsprop",loss="binary_crossentropy", 
             metrics=["accuracy"])

cnn3.fit(X_train, y_train, epochs = 5, batch_size = 20)

cnn3_elapsed_time = time.time() - cnn3_start_time

Epoch 1/5
1600/1600 [==============================] - 62s 39ms/step - loss: 8.0430 - acc: 0.4963
Epoch 2/5
1600/1600 [==============================] - 61s 38ms/step - loss: 8.0389 - acc: 0.5013
Epoch 3/5
1600/1600 [==============================] - 74s 46ms/step - loss: 8.0389 - acc: 0.5013
Epoch 4/5
1600/1600 [==============================] - 84s 52ms/step - loss: 8.0389 - acc: 0.5013
Epoch 5/5
1600/1600 [==============================] - 92s 58ms/step - loss: 8.0389 - acc: 0.5013


In [68]:
print (round(cnn3_elapsed_time,2), "seconds")

373.81 seconds


#####  Classifier 4 Specs
* Four convolution layers
* 16 nodes per layer
* ReLU activation function except for the output layer (Sigmoid)
* Kernel size = 3 (3x3 filter matrix) 

In [69]:
cnn4_start_time =time.time()

cnn4 = Sequential()
cnn4.add(Conv2D(16, kernel_size = 3, activation ="relu", input_shape=(64,64,1)))
cnn4.add(Conv2D(16, kernel_size =3, activation ="relu"))
cnn4.add(Conv2D(16, kernel_size =3, activation ="relu"))
cnn4.add(Conv2D(16, kernel_size =3, activation ="relu"))
#Connection between convolution and dense layer
cnn4.add(Flatten())
cnn4.add(Dense(1, activation = "sigmoid"))

cnn4.compile(optimizer = "rmsprop",loss="binary_crossentropy", 
             metrics=["accuracy"])

cnn4.fit(X_train, y_train, epochs = 5, batch_size = 20)

cnn4_elapsed_time = time.time() - cnn4_start_time

Epoch 1/5
1600/1600 [==============================] - 29s 18ms/step - loss: 7.9444 - acc: 0.4963
Epoch 2/5
1600/1600 [==============================] - 28s 17ms/step - loss: 7.9911 - acc: 0.4988
Epoch 3/5
1600/1600 [==============================] - 28s 17ms/step - loss: 7.9911 - acc: 0.4988
Epoch 4/5
1600/1600 [==============================] - 32s 20ms/step - loss: 7.9911 - acc: 0.4988
Epoch 5/5
1600/1600 [==============================] - 37s 23ms/step - loss: 7.9911 - acc: 0.4988


In [70]:
print (round(cnn4_elapsed_time,2), "seconds")

154.48 seconds


##### Key Takeaways
* This section was where we trained each neural network using the training data set 
* Specifically, we trained four different CNNs
* The only difference between each CNN was the # of convolution layers and the # of nodes within each layer 
* Classifier 1: 2 layers and 32 nodes
* Classifier 2: 2 layers and 16 nodes
* Classifier 3: 4 layers and 32 nodes
* Classifier 4: 4 layers and 16 nodes 
* We noted that adding additional layers significantly increased the development time 
* Final observations will be communicated after model validation

### Section 4 - Model Validation

#### This section will calculate the classification accuracy for the training and test sets for each model. We will apply the evaluate method on each model and use the training or test sets with the same batch size as parameters. 

In [71]:
#Classifier 1
#Evaluate model on test data set
score_test1 = cnn1.evaluate(X_test, y_test, batch_size=20)

400/400 [==============================] - 3s 8ms/step


In [79]:
#Evaluate model on training data set 
score_train1 = cnn1.evaluate(X_train, y_train, batch_size =20)

1600/1600 [==============================] - 9s 6ms/step


In [83]:
#This prints out the loss value and accuracy, respectively 
#Printing out the result from training and test set, respectively 
print (score_train1, score_test1)

[8.038900166749954, 0.5012500012293458] [8.139638304710388, 0.49500000178813935]


In [84]:
#Classifier 2
#Evaluate model on training data set
score_train2 = cnn2.evaluate(X_train, y_train, batch_size=20)

1600/1600 [==============================] - 5s 3ms/step


In [85]:
#Evaluate model on test data set 
score_test2 = cnn2.evaluate(X_test, y_test, batch_size =20)

400/400 [==============================] - 1s 3ms/step


In [86]:
#This prints out the loss value and accuracy, respectively 
#Printing out the result from training and test set, respectively 
print (score_train2, score_test2)

[8.038900166749954, 0.5012500012293458] [8.139638304710388, 0.49500000178813935]


In [87]:
#Classifier 3
#Evaluate model on training data set
score_train3 = cnn3.evaluate(X_train, y_train, batch_size=20)

1600/1600 [==============================] - 23s 15ms/step


In [88]:
#Evaluate model on test data set 
score_test3 = cnn3.evaluate(X_test, y_test, batch_size =20)

400/400 [==============================] - 5s 13ms/step


In [89]:
#This prints out the loss value and accuracy, respectively 
#Printing out the result from training and test set, respectively 
print (score_train3, score_test3)

[8.038900166749954, 0.5012500012293458] [8.139638304710388, 0.49500000178813935]


In [90]:
#Classifier 4
#Evaluate model on training data set
score_train4 = cnn4.evaluate(X_train, y_train, batch_size=20)

1600/1600 [==============================] - 12s 7ms/step


In [91]:
#Evaluate model on test data set 
score_test4 = cnn4.evaluate(X_test, y_test, batch_size =20)

400/400 [==============================] - 3s 6ms/step


In [92]:
#This prints out the loss value and accuracy, respectively 
#Printing out the result from training and test set, respectively 
print (score_train4, score_test4)

[7.991120874881744, 0.49874999914318324] [7.89148097038269, 0.505000002682209]


### Key Takeaways
* The results from this exercise were disappointing 
* All of the models had relatively the same accuracy performance on the test set -- approximately 50%, which is the equivalent to a coin flip in terms of predicting whether an image is a dog or cat
* The model with the best performance was Classifier 4, which had an accuracy score of 50.5% on the test set 
* In terms of studying the impact of the benchmark experiment, there really isn't much to conclude since the results were very similar. In other words, there was no observable difference between the different factor inputs of layers and nodes/layer 
* If development time were important, we would choose Classifier 2 because its accuracy score was 49.5% but it had the shortest time 
* Overall, further experimentation is required to improve the accuracy scores. As noted, hyperparameters play an important part in developing a CNN 

![alt text](HW7.jpg "Table1")

### Section 5 - Benchmark Experiment Results and Final Recommendation

Based on past research, we know that convolutional neural networks work best for image classification, speech recognition and natural language processing. Therefore, our final recommendation would be for the website provider to utilize a convolutional nueral network to help them classify images provided by their users. Unfortunately, the results of this experiment do not provide evidence for such a recommendation given the poor performance as measured by binary classification accuracy. Reflecting back on our results, we did not use multiple connected layers or make any augmentations to the data that was provided. In other words, further hyperparameter tuning seems to be an option that should be explored to improve the accuracy of these models. Therefore, because of the results, we cannot make an appropriate recommendation on which type of CNN to use. As mentioned, the lack of data augmentation may have played a part in our results. It would seem that using images with a larger shape (greater than 64 x 64) and on the RGB scale work best for this data set. 